rfcs/0000-central-station.md

```diff
@@ -0,0 +1,507 @@
+  - Terminal departure times of trips, with dispatcher-entered "offsets" in minutes
+- "Device" information, like state of switches and terminal bells, and hold lights
+  - We mostly don't use any of this.
+
+RTR uses this raw data to keep track of where the known subway vehicles are. It attempts to match
+current revenue OCS trips to a plausible GTFS trip, creating an "ADDED" trip if none match. It
+publishes this data as VehiclePositions.
+
+RTR also generates predictions for most revenue vehicles' current trips and some subsequent trips.
+It publishes this data as TripUpdates. Unmatched GTFS trips it publishes as CANCELED.
+
+In the future, we may get raw GPS updates from all light rail vehicles once every 5 seconds, both
+in and out of revenue service.
+
+The RTR data is GTFS-rt-esque; The TripUpdates also includes our "boarding_status" extension for,
+e.g., "Stopped 4 stops away" for Realtime Signs, a passthrough_time field for non-revenue trains
+for Realtime Signs to play the "please stand back from the yellow line" message, route_pattern_id
+for ADDED trips. The VehiclePositions include "consist" information.
+
+Realtime Signs consumes the RTR `_enhanced.json` directly, rather than via the public API for
+those extra fields above. Glides consumes it directly, and would ideally like the VehiclePositions
+of all light rail trains, in service or not.
+
+### With CentralStation
+
+RTR will change to behave more like how we want Swiftly to behave.
+
+It will consume VehiclePositions data from CentralStation, and importantly, those vehicle
+positions will have a GTFS trip tagged. This will simplify RTR as a large part of the message
+handling logic is about deciding what trip a vehicle is on.
+
+It will also consume the ServiceChanges feed, which is how CentralStation will convey the TSCH
+offsets it receives.
+
+As a bonus, by relying on GTFS trips and ServiceChanges, RTR won't have to rely on handcrafted
+patterns, and should be able to predict, e.g., new stations or closed stations automatically.
+
+Question: how do you make reverse predictions this way? Is there something in this data model that
+allows us to convey to RTR the _subsequent_ trip for a vehicle?
+
+## Busloc
+
+### Current description
+
+Busloc primarily keeps track of the whereabouts of MBTA buses and publishes a few feeds related to
+them. Busloc runs on the MBTA network because one of the sources of its data, the TransitMaster
+DB, is only accessible from there.
+
+The majority of what Busloc deals with is "normal" bus data. These are buses logged on and running
+their expected, planned routes. Busloc gets this data from TransitMaster, via its hacky internal
+spelunking of the TransitMaster DB, and via Samsara, a cloud service that has more frequent GPS
+updates. Busloc combines these feeds, taking the latest lat/long from either source for every
+vehicle. These logged-on vehicles have their GTFS trip directly in the data, since TransitMaster
```

Comment:
One thing to keep in mind is that whenever a bus has a GTFS trip ID, that's because something else (TM, Swiftly, etc.) has already done its own trip matching logic. Operators only log in with run ID, so if you get trip ID from something else you're already relying on some other service's algorithm.